Add Word Break problem

ykushch opened the pull request:
Solution for the following problems:
- Edit Distance problem (only naive)
- Longest Common Subsequence (only naive)
- Break Word problem (recursive & dynamic programming approach)

ykushch committed:
Add Word Break problem

vseloved commented on tasks/ykushch/src/main/java/com/ykushch/prjalgo2/task1/WordBreak.java:
```diff
@@ -0,0 +1,89 @@
+                continue;
+            }
+
+            for (String wordFromDict : dict) {
```
this means that we'll be doing O(n \* dict.size) operations where dict.size >> n (usually). Not good. We can do in O(n)

vseloved commented on tasks/ykushch/src/main/java/com/ykushch/prjalgo2/task4/DistanceRecursive.java:
```diff
@@ -0,0 +1,62 @@
+    private static boolean areLastCharsEqual(char srcLastChar, char destLastChar) {
+        return srcLastChar == destLastChar;
+    }
+}
```
не хватает вывода списка букв для выравненных строк

ykushch committed:
Merge remote-tracking branch 'upstream/master'

ykushch committed:
Fix inefficient approach and add new solutions

There is an inefficient approach with dictionary. Now it searches for words manually adding one symbol at a time.
Add solutions for:
* Binary Heap
* Minimum Edit Distance

ykushch commented:
@vseloved please, review changes with updates.

ykushch committed:
Add max flow solution

vseloved commented on tasks/ykushch/src/main/java/com/ykushch/prjalgo2/task1/WordBreak.java:
```diff
@@ -86,4 +86,51 @@ public String breakWord(String s, Set<String> dict) {
+    public String breakWordWithLargeDictionary(String s, Set<String> dict) {
+        List<String> resultList = new ArrayList<>();
+        List<String> memoList = new ArrayList<>();
+        isSentenceBrokenIntoWords(s, dict, resultList, memoList);
```
тут бы не помешала еще проверка на успешное разбитие строки

ykushch replied:
она вроде бы по коду ниже есть. Или я что-то не так понял?
